lib/membrane/core/bin/pad_controller.ex

```diff
@@ -0,0 +1,260 @@
+    state.pads.data
+    |> Map.values()
+    |> Enum.filter(&(&1.spec_ref == spec_ref))
+    |> Enum.reduce(state, fn pad_data, state ->
+      if pad_data.link_id do
+        Message.send(state.parent_pid, :link_response, pad_data.link_id)
+        state
+      else
+        Membrane.Core.Child.PadModel.set_data!(
+          state,
+          pad_data.ref,
+          :response_received?,
+          true
+        )
+      end
+    end)
+  end
+
+  @doc """
+  Returns true if all pads of given `spec_ref` are linked, false otherwise.
+  """
+  @spec all_pads_linked?(ChildLifeController.spec_ref_t(), State.t()) :: boolean()
+  def all_pads_linked?(spec_ref, state) do
+    state.pads.data
+    |> Map.values()
+    |> Enum.filter(&(&1.spec_ref == spec_ref))
+    |> Enum.all?(& &1.linked?)
+  end
+
+  @doc """
+  Verifies linked pad and proxies the message to the proper child.
+  """
+  @spec handle_link(
+          Pad.direction_t(),
+          LinkParser.raw_endpoint_t(),
+          LinkParser.raw_endpoint_t(),
+          PadModel.pad_info_t() | nil,
+          map,
+          Core.Bin.State.t()
+        ) :: Type.stateful_try_t(PadModel.pad_info_t(), Core.Bin.State.t())
+  def handle_link(direction, this, other, other_info, link_metadata, state) do
+    pad_data = PadModel.get_data!(state, this.pad_ref)
+    %{spec_ref: spec_ref, endpoint: endpoint} = pad_data
+
+    pad_props =
+      Map.merge(this.pad_props, endpoint.pad_props, fn key, external_value, internal_value ->
+        if key in [
+             :demand_excess_factor,
+             :min_demand_factor,
+             :auto_demand_size_factor,
+             :toilet_capacity_factor
+           ] do
+          external_value || internal_value
+        else
+          internal_value
+        end
+      end)
+
+    endpoint = %{endpoint | pad_props: pad_props}
+
+    :ok =
+      Child.PadController.validate_pad_mode!(
+        {this.pad_ref, pad_data},
+        {other.pad_ref, other_info}
+      )
+
+    reply =
+      Message.call(endpoint.pid, :handle_link, [
+        direction,
+        endpoint,
+        other,
+        other_info,
+        link_metadata
+      ])
+
+    state = PadModel.set_data!(state, this.pad_ref, :linked?, true)
+    state = PadModel.set_data!(state, this.pad_ref, :endpoint, endpoint)
+    state = ChildLifeController.LinkHandler.proceed_spec_linking(spec_ref, state)
+    {reply, state}
+  end
+
+  @doc """
+  Handles situation where pad has been unlinked (e.g. when connected element has been removed from pipline)
+  """
+  @spec handle_unlink(Pad.ref_t(), Core.Bin.State.t()) :: Type.stateful_try_t(Core.Bin.State.t())
+  def handle_unlink(pad_ref, state) do
+    with {:ok, state} <- handle_pad_removed(pad_ref, state) do
+      PadModel.delete_data(state, pad_ref)
+    end
+  end
+
+  @spec handle_pad_added(Pad.ref_t(), Core.Bin.State.t()) ::
+          Type.stateful_try_t(Core.Bin.State.t())
+  defp handle_pad_added(ref, state) do
```

Comment:
I observed that usually when a function does something for some conditions and is a NOP for others (in this case if the `availability` is not `:dynamic`) it is prefixed with a `maybe_`
```suggestion
  defp maybe_handle_pad_added(ref, state) do
```